### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by AndriiKlymchuk**
Hello. I'm gonna use this lib in production, but while testing I found several bugs. After reading the code I spot multiple areas for performance and readability improvements. They were my main focus during refactoring as well as bug fixing.
Although there are a lot of changes, lib interface remains as it was before and all old tests + some new are pass.

Short summary of changes:
- Fix incorrect next rank calculation when step is grater then 122
- Fix incorrect middle value calculation when both ranks are divisible by two with reminder
- Optimize and simplify code
- Run mix.format
- Rename some terms for better clarity, e.g. digits -> numbers
- Add dialyzer specs
- Fix typos

### CODE REVIEW

**Comment by AndriiKlymchuk** [comment]
@fastindian84 please let me know what do you think about this PR and if you willing to merge it.

**Comment by fastindian84** [comment]
> @fastindian84 please let me know what do you think about this PR and if you willing to merge it.

Thx, I'll take a look this week

File: lib/lexorank_ex.ex [hunk 1]
```diff
@@ -1,100 +1,112 @@
...
+  The result is a middle rank for NumeralSystem with radix == 62
   """
+  @spec middle(non_neg_integer()) :: String.t()
   def middle(division) when is_integer(division) do
-    import LexorankEx.NumerialSystem, only: [min: 1, max: 1]
-
-    between(min(division), max(division))
+    between(NumeralSystem.min(division), NumeralSystem.max(division))
   end
-  def middle(_), do: raise LexorankEx.Error.exception("Please, provide division number")
```
Comment (fastindian84) [inline lib/lexorank_ex.ex:60]: Why did you remove this line?

File: lib/lexorank_ex.ex [hunk 2]
```diff
@@ -1,100 +1,112 @@
...
+
   def between(left, right) do
     [min, max] = Enum.sort([left, right])
+    min_numbers = Utils.to_numbers(min)
+    max_numbers = Utils.to_numbers(max)
+    {min_numbers, max_numbers} = Utils.equalize_length(min_numbers, max_numbers)
+
+    Add.call(min_numbers, max_numbers)
+    |> Divide.call()
+    # Drop extra undesired item which may be introduced by Add.call/2, e.g.
```
Comment (fastindian84) [inline lib/lexorank_ex.ex]: Please remove undesired , it sounds more like a bug produced by the system, and the system was designed to do so.

File: lib/lexorank_ex.ex [hunk 3]
```diff
@@ -1,100 +1,112 @@
...
-    sum = Add.call(min_half, max_half)
-
-    case Compare.between?(min_digits, sum, max_digits) do
-      true -> Coersion.to_chars(sum)
-      false ->
-        (Fraction.call(reminder) ++ sum)
-        |> Coersion.to_chars()
-    end
+  defp resolve_clash(min, min) do
+    [div(NumeralSystem.radix(), 2) | min]
```
Comment (fastindian84) [inline lib/lexorank_ex.ex]: `NumeralSystem.middle_index`

File: test/lexorank_ex_test.exs [hunk 4]
```diff
@@ -49,17 +51,22 @@ defmodule LexorankExTest do
...
+    assert LexorankEx.between("3", "5") == "4"
   end
 
   test "#distance/2" do
-    assert LexorankEx.distance("V0000007", "V000000F") == 8
-    assert LexorankEx.distance("00", "zz") == 3843
-    assert LexorankEx.distance("000", "zzz") == 238327
-    assert LexorankEx.distance("aa", "bb") == 63
-    assert LexorankEx.distance("00", "010") == 62
+    # assert LexorankEx.distance("V0000007", "V000000F") == 8
```
Comment (fastindian84) [inline test/lexorank_ex_test.exs]: Why did you comment these?

File: test/bucket_test.exs [hunk 5]
```diff
@@ -60,24 +62,24 @@ defmodule Bucket.BucketTest do
     assert Bucket.between("1|a", "1|c") == "1|b"
     assert Bucket.between("1|0", "1|z") == "1|U"
     assert Bucket.between("2|AA", "2|AB") == "2|AAV"
-    assert Bucket.between("2|aaaaz", "2|zzzzz") == "2|nIIIT"
+    assert Bucket.between("2|aaaaz", "2|zzzzz") == "2|nIIIU"
```
Comment (fastindian84) [inline test/bucket_test.exs:65]: why did this one change?

**Review by fastindian84 — COMMENTED** [review]
Thx, sorry for long reply, most comments are stylish, the only one related to calc is 

Why this has changed:`Bucket.between("2|aaaaz", "2|zzzzz") == "2|nIIIT"`

File: test/bucket_test.exs (hunk 5 above)
Comment (AndriiKlymchuk) [reply test/bucket_test.exs:65]: Because there was a bug which leads to incorrect middle value calculation in some cases which I fixed. Lets verify it:
```
LexorankEx.Bucket.distance("2|aaaaz", "2|nIIIT") == 187732124
LexorankEx.Bucket.distance("2|zzzzz", "2|nIIIT") == 187732126
```
As you see distance between lowest and smallest values is not equal, moreover its difference is 2
Now, lets check correct value:
```
LexorankEx.Bucket.distance("2|aaaaz", "2|nIIIU") == 187732125
LexorankEx.Bucket.distance("2|zzzzz", "2|nIIIU") == 187732125
```
Distance is equal, it means that middle value is calculated correctly.

File: test/lexorank_ex_test.exs (hunk 4 above)
Comment (AndriiKlymchuk) [reply test/lexorank_ex_test.exs]: Accidentally, fixed

File: lib/lexorank_ex.ex (hunk 3 above)
Comment (AndriiKlymchuk) [reply lib/lexorank_ex.ex]: Done

File: lib/lexorank_ex.ex (hunk 2 above)
Comment (AndriiKlymchuk) [reply lib/lexorank_ex.ex]: Agree, done

File: lib/lexorank_ex.ex (hunk 1 above)
Comment (AndriiKlymchuk) [reply lib/lexorank_ex.ex:60]: Because it's basically just type check. We do it only in this function. I've added dialyzer specs which covers all type errors, so we don't need to write this kind of guard code.

**Comment by AndriiKlymchuk** [comment]
@fastindian84 fixed and answered all your comments. Please look again.

File: test/bucket_test.exs (hunk 5 above)
Comment (fastindian84) [reply test/bucket_test.exs:65]: oh, I see, thx!

**Review by fastindian84 — APPROVED** [review]

**Comment by fastindian84** [comment]
Thank you, merged
